grails-app/services/org/pih/warehouse/tableroapi/NumberDataService.groovy

```diff
@@ -13,22 +13,35 @@ class NumberDataService {
         Date tomorrow = new Date() + 1;
         tomorrow.clearTime();
 
-        def binLocations = InventorySnapshot.executeQuery('select count(*) from InventorySnapshot i where i.location=:location and i.date = :tomorrow and i.quantityOnHand > 0', ['location': location, 'tomorrow': tomorrow]);
-        def shipments = Requisition.executeQuery("""select count(*) from Requisition r where r.origin = :location and r.status <> 'ISSUED' and r.createdBy = :user""", 
+        def binLocations = InventorySnapshot.executeQuery("select count(*) from InventorySnapshot i where i.location=:location and i.date = :tomorrow and i.quantityOnHand > 0", ['location': location, 'tomorrow': tomorrow]);
+
+        def shipments = Requisition.executeQuery("select count(*) from Requisition r where r.origin = :location and r.status <> 'ISSUED' and r.createdBy = :user", 
         ['location': location, 'user': user]);
+        
         def incompletePutaways = Order.executeQuery("select count(o.id) from Order o where o.orderTypeCode = 'TRANSFER_ORDER' AND o.status = 'PENDING' AND o.orderedBy = :user AND o.destination = :location", ['user':user, 'location': location]);
 
+        def receivingBin = InventorySnapshot.executeQuery("""
+            SELECT COUNT(distinct i.product) from InventorySnapshot i 
+            LEFT JOIN i.location l 
+            WHERE l = :location AND i.quantityOnHand > 0 
+            AND i.date = :tomorrow AND l.locationType = 'ff8081816482352b01648249e8cc0001'""",
```

Comment:
This should not be hard-coded. Can you explain what you're trying to do here?

Reply:
This is the calculation we got for this indicator. Should we define a variable with the selected locationType and keep it as a parameter?

Reply:
Ah, ok. Sigh. I thought Receiving had it's own location type code, but I guess it's INTERNAL just like Bin Location. For now let's make this a constant. In fact, any time you need to hard-code a primary key, please create a constant. If it becomes necessary, we can make these configurable in Config.groovy. Eventually, it might be good to make some of thees indicators more generic so we can configure them. This one is specific to the receiving workflow, but I could see someone wanting to clone this indicator with their own custom location type (i.e. Quarantine).
```
l.locationType = '${Constants.RECEIVING_LOCATION_TYPE_ID}'
```
In Constants.groovy declare RECEIVING_LOCATION_TYPE_ID under the other two location types.
```
    // direct references to locations by primary key
    static final String WAREHOUSE_LOCATION_TYPE_ID = "2"
    static final String SUPPLIER_LOCATION_TYPE_ID = "4"
    static final String RECEIVING_LOCATION_TYPE_ID = "ff8081816482352b01648249e8cc0001"
```

Reply:
Also, does this actually work? 
```
AND l.locationType = '<pk>'
```
I figured it would need to be `l.locationType.id` but if it works that's a nice shorthand to know. 

In order for this to work, Hibernate needs to be smart enough to know that either the `l.locationType` should reference the primary key OR it joins/loads the LocationType with ```id = '<pk>'``` and then . I'm not sure which it does but I'm curious a) that it works and b) doesn't generate extra queries / joins.

My understanding is that ```l.locationType.id = '<pk>'``` would not require the join between Location and LocationType so it would be more efficient. 
> The special property (lowercase) id can be used to reference the unique identifier of an object. See Section 15.5, “Referring to identifier property” for more information.
> ```from Cat as cat where cat.id = 123```
> ```from Cat as cat where cat.mate.id = 69```
> The second query is efficient and does not require a table join.

So if you have time, let me know if Hibernate is smart enough to not require a join when it evaluates 
```l.locationType = '<pk>'```.

It's not a big deal for this particular indicator, but I want to make sure we know what's happening behind the scenes and using the most efficient HQL possible.

Reply:
When debugging the intitial request with `l.locationType = ...` the actual query is this:
```
select count(distinct inventorys0_.product_id) as col_0_0_
from inventory_snapshot inventorys0_
left outer join location location1_ on inventorys0_.location_id=location1_.id, product product2_
where inventorys0_.product_id=product2_.id and location1_.id=?
and inventorys0_.quantity_on_hand>0 and inventorys0_.date=?
and location1_.location_type_id='ff8081816482352b01648249e8cc0001'
```
so it looks like Hibernate does not generate an extra join, but simply checks the value of the foreign key on the Location table. It's the exact same query with `l.locationType.id = ...`.

On the other hand the Product inside the select is joined when we do `count(distinct i.product)` (above) but not with `count(distinct i.product.id)`:
```
select count(distinct inventorys0_.product_id) as col_0_0_
from inventory_snapshot inventorys0_
left outer join location location1_ on inventorys0_.location_id=location1_.id
where location1_.id=? and inventorys0_.quantity_on_hand>0 and inventorys0_.date=?
and location1_.location_type_id='ff8081816482352b01648249e8cc0001'
```
We'll put the `.id` in both, but no extra joins are generated to evaluate the foreign key in the where clause.